File: 프로그래머스/C++/LV3_여행경로.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;

bool check[10001];


vector <string> answer;

bool dfs(string start, int cnt, int end_cnt, vector<vector<string>> tickets, vector<string> v) {

	if (cnt == end_cnt) {
		for (int i = 0; i < v.size(); i++) {
			answer.push_back(v[i]);
		}
		return true;
	}

	for (int i = 0; i < tickets.size(); i++) {
		if (start == tickets[i][0] && !check[i]) {
			check[i] = true;
			v.push_back(tickets[i][1]);
			bool result = dfs(tickets[i][1], cnt + 1, end_cnt, tickets, v);
			if (result)
				return true;
			check[i] = false;
			v.pop_back();
		}
	}

	return false;

}


vector<string> solution(vector<vector<string>> tickets) {

	sort(tickets.begin(), tickets.end()); //알파벳 순서대로 정렬
	string start = "ICN";

	answer.push_back(start);
	int cnt = 0;
	int end_cnt = tickets.size();

	vector <string> v;
	dfs(start, cnt, end_cnt, tickets, v);

	return answer;
}

File: 프로그래머스/JAVA/LV3_여행경로.java
import java.util.*;
class Solution {
    static String[] answer;
    static ArrayList<String> port;
    static boolean[] visited;
    public String[] solution(String[][] tickets) {
        answer = new String[tickets.length+1];
        port = new ArrayList<>();
        visited = new boolean[tickets.length];
        
        Arrays.sort(tickets, new Comparator<String[]>(){
            @Override
            public int compare(String[] s1, String[] s2){
                if(s1[0].equals(s2[0])){
                    return s1[1].compareTo(s2[1]);
                }else
                    return s1[0].compareTo(s2[0]);
            }
        });
        
        
        
        dfs("ICN", 0, tickets);
        
        
        return answer;
    }
    
    public void dfs(String now, int n, String[][] tickets){
        if(answer[0] != null) return;
        
        if(n == tickets.length){
            port.add(now);
            answer = new String[n+1];
            
            for(int i = 0; i<port.size(); i++){
                answer[i] = port.get(i);
            }
            return;
        }
        
        for(int i = 0; i<tickets.length; i++){
            if(visited[i]){
                continue;
            }
            if(tickets[i][0].equals(now)){
                port.add(now);
                visited[i] = true;
                
                dfs(tickets[i][1], n+1, tickets);
                visited[i] = false;
                port.remove(port.size()-1);
            }
        }
    }
}
